File: Laba2/OP2.2Python/Func.py
from datetime import date
import pickle
from autoBase import AutoBase
from date import Date

def inpt(flag): 
    k = '+'
    if flag: File = open("AutoBase.txt", "wb") # Відкриття бінарного файла для запису
    else: File = open("AutoBase.txt", "ab") # Відкриття бінарного файла для дозапису
    while k == "+":
        Name = input("Enter name: ")
        Data = input("Enter Release date with point '10.10.2010': ")
        DateR = Date(int(Data[:Data.find('.')]), int(Data[Data.find('.') + 1: Data.rfind('.')]), int(Data[Data.rfind('.') + 1:]))
        Data = input("Enter Sale Release date with point '10.10.2010': ")
        DateSR = Date(int(Data[:Data.find('.')]), int(Data[Data.find('.') + 1: Data.rfind('.')]), int(Data[Data.rfind('.') + 1:]))
        auto = AutoBase(Name, DateR, DateSR)
        k = input("\nEnter '+' in case you want to contnue: ")
        pickle.dump(auto, File)
    File.close()
    
def ReadF(name): # Читання з файла у список
    File = open(name, "rb") # Відкриття файла для читання
    list =[]
    flag = True 
    while flag:
        try:
            list.append(pickle.load(File))
        except EOFError: # Поки не дійшло до кінця файла
            flag = False
    File.close()
    return list

def outpt(name): # Вивід файла
    list = ReadF(name)
    for auto in list:
        print("Name: {:<15} Release Date: {!s:0>2}.{!s:0>2}.{!s:<9} Sale Release Date: {!s:0>2}.{!s:0>2}.{!s}".format(auto.Name, auto.DateR.day, auto.DateR.mounth, auto.DateR.year, 
                        auto.DateSR.day, auto.DateSR.mounth, auto.DateSR.year))

def CreateList(): 
    date_now = date.today() # Зчитування поточної дати
    File = open("AutoList.txt", "wb") # Відкриття бінарного файлу для запису
    list = ReadF("AutoBase.txt")
    flag = True
    d1 = date_now.day + date_now.month*31 + date_now.year*365
    for auto in list:
        d2 = auto.DateSR.day + auto.DateSR.mounth*31 + auto.DateSR.year*365
        if d2 + 31 >= d1: # За останній місяць
            flag = False
            pickle.dump(auto, File)
    File.close()
    print("\nCars on Sale in last mounth: ")
    if flag: print("No cars!")
    else: outpt("AutoList.txt")

def UseedAuto():
    print("\nSaled after >1 year after Release: ")
    File = open("AutoBase.txt", "rb")  # Відкриття бінарного файлу для читання
    flag = True
    list = ReadF("AutoBase.txt") # Читання з файла у список
    for auto in list:
        d1 = auto.DateSR.day + auto.DateSR.mounth*31 + auto.DateSR.year*365
        d2 = auto.DateR.day + auto.DateR.mounth*31 + auto.DateR.year*365
        if d2 + 365 < d1:
            flag = False
            print("Name: {:<15} Release Date: {!s:0>2}.{!s:0>2}.{!s:<9} Sale Release Date: {!s:0>2}.{!s:0>2}.{!s}".format(auto.Name, auto.DateR.day, auto.DateR.mounth, auto.DateR.year, 
                        auto.DateSR.day, auto.DateSR.mounth, auto.DateSR.year))
    File.close()
    if flag: print("No cars!")

            

File: Laba2/OP2.2/Func.cpp
#pragma warning(disable:4996)
#define Sal a.SaleRelease
#define Rel a.Release
#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>
#include "OP2.2.h"
using namespace std;

void input(char cont) {
	char k; // Зберігає символ, за допомогою якого визначається чи продовжувати запис
	AutoBase a; // Змінна, що зберігає інформація про машину
	string Read; // Змінна для читання з консолі
	ofstream In; 
	if (cont == '+') { In.open("AutoBase.txt", ios::binary | ios::app); cin.ignore(); }// Відкриття файла для дозапису
	else In.open("AutoBase.txt", ios::binary); // Відкриття файла для запису
	do {
		cout << "Enter name of car: "; getline(cin, Read); // Ім'я машини
		strcpy(a.Name, data(Read));
		cout << "Enter Release Date with a point '10.10.2010': "; getline(cin, Read); // Дата виготовлення
		a = Divide(a, Read, 0); // Розділення дати у форматі string на змінні формату int
		cout << "Enter Sale Release Date with a point '10.10.2010': "; getline(cin, Read); // Дата продажу
		a = Divide(a, Read, 1); // Розділення дати у форматі string на змінні формату int
		cout << "\nEnter '+' in case you want to contnue: "; cin >> k; // Чи хоче користувач продовжувати
		In.write((char*)&a, sizeof(AutoBase)); // Запис у файл даних про машину
		cin.ignore();
	} while (k == '+');
	In.close();
}

AutoBase Divide(AutoBase a, string Read, bool b) {
	int pos1 = Read.find('.'); // Позиція першої крапки
	int pos2 = Read.rfind('.'); // Позиція другої крапки
	if (!b) {
		Rel.day = stoi(Read.substr(0, pos1)); // Запис дня у змінну
		Rel.mounth = stoi(Read.substr(pos1 + 1, pos2)); // Запис місяця у змінну
		Rel.year = stoi(Read.substr(pos2 + 1)); // Запис року у змінну
	}
	else {
		Sal.day = stoi(Read.substr(0, pos1)); // Запис дня у змінну
		Sal.mounth = stoi(Read.substr(pos1 + 1, pos2)); // Запис місяця у змінну
		Sal.year = stoi(Read.substr(pos2 + 1)); // Запис року у змінну
	}
	return a;
}

void CreateList() {
	string date = datta(); // Поточна дата
	ifstream Out("Autobase.txt", ios::binary); // Відкриття файла для читання
	ofstream In("OneMounth.txt", ios::binary); // Відкриття файла для запису
	AutoBase a; // Змінна, що зберігає інформація про машину
	bool flag = true; // Наявність машин, що відповідають умовам
	int day = stoi(date.substr(0, date.find('.'))); // Поточний день
	int mounth = stoi(date.substr(date.find('.') + 1, date.rfind('.'))); // Поточний місяць
	int year = stoi(date.substr(date.rfind('.') + 1)); // Поточний рік
	int d1 = day + mounth * 31 + year * 365; // Поточна дата у днях
	int d2; // Дата продажу у днях
	while (Out.read((char*)&a, sizeof(AutoBase))) { // Читання з файлу
		d2 = Sal.day + Sal.mounth * 31 + Sal.year * 365;
		if (d2 + 31 >= d1) { // Мащини, що були продані в останній місяць
			In.write((char*)&a, sizeof(AutoBase)); // Запис у файл
			flag = false;
		}
	}
	In.close();
	Out.close();
	cout << "\nCars on Sale in last mounth: \n";
	if (flag) cout << "No cars.\n"; // Якщо таких машин нема
	else output("OneMounth.txt");// Вивід зміста файла
}

string datta() { // Поточна дата
	const time_t tm = time(nullptr);
	char buf[64];
	strftime(buf, sizeof(buf), "%d.%m.%Y", localtime(&tm));
	string date = buf; 
	return date;
}

void output(string name) {
	ifstream Out(name, ios::binary); // Відкриття файла для читання
	AutoBase a; // Змінна, що зберігає інформація про машину
	while (Out.read((char*)&a, sizeof(AutoBase))) // Читання з файла
		printf("Name: %-15s Release date: %02i.%02i.%-9i Sale Release date: %02i.%.02i.%-9i\n",
			a.Name, Rel.day, Rel.mounth, Rel.year, Sal.day, Sal.mounth, Sal.year);
	Out.close();
}

void UseedAuto() {
	cout << "\nSaled after >1 year after Release: \n";
	ifstream Out("Autobase.txt", ios::binary); // Відкриття файла для читання
	AutoBase a; // Змінна, що зберігає нформація про машину
	bool flag = true; // Наявність машин, що відповідають умовам
	int d1, d2; // Дати у днях
	while (Out.read((char*)&a, sizeof(AutoBase))) {// Читання з файла
		d1 = Sal.day + Sal.mounth * 31 + Sal.year * 365; // Дата виготовлення
		d2 = Rel.day + Rel.mounth * 31 + Rel.year * 365; // Дата продажу
		if (d2 + 365 < d1) { // Машини, що були продані більше ніж через рік після створення
			printf("Name: %-15s Release date: %02i.%02i.%-9i Sale Release date: %02i.%.02i.%-9i\n",
				a.Name, Rel.day, Rel.mounth, Rel.year, Sal.day, Sal.mounth, Sal.year);
			flag = false;
		}
	}
	if (flag) cout << "No cars.\n"; // Якщо таких машин нема
	Out.close();
}